source/val/function.cpp

```diff
@@ -371,4 +373,45 @@ Construct& Function::FindConstructForEntryBlock(const BasicBlock* entry_block) {
   return *construct_ptr;
 }
 
+int Function::GetBlockDepth(BasicBlock* bb) {
+  if (!bb) {
+    return 0;
+  }
+  // Only calculate the depth if it's not already calculated.
+  // This function uses memoization to avoid duplicate CFG depth calculations.
+  if (block_depth_.find(bb) == block_depth_.end()) {
+    if (bb->is_type(kBlockTypeMerge)) {
+      // If this is a merge block, its depth is equal to the block before
+      // branching.
+      BasicBlock* header = merge_block_header_[bb];
+      assert(header);
+      block_depth_[bb] = GetBlockDepth(header);
+    } else if (bb->is_type(kBlockTypeContinue)) {
+      // The depth of the continue block entry point is 1 + loop header depth.
+      Construct* continue_construct = entry_block_to_construct_[bb];
+      assert(continue_construct);
+      // Continue construct has only 1 corresponding construct (loop header).
+      Construct* loop_construct =
+          continue_construct->corresponding_constructs()[0];
+      assert(loop_construct);
+      BasicBlock* loop_header = loop_construct->entry_block();
+      block_depth_[bb] = 1 + GetBlockDepth(loop_header);
+    } else {
+      BasicBlock* bb_dom = bb->immediate_dominator();
+      if (!bb_dom || bb_dom == bb) {
```

Comment:
I believe bb_dom can never be null.  So some might get confused about the !bb_dom condition.  But you could argue it's defensive.
Also, perhaps the doc for BasicBlock::immediate_dominator should say its result is never null. 

ok to keep.